ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpImplTreeBuilder.fs

```diff
@@ -625,10 +637,12 @@ type FSharpImplTreeBuilder(lexer, document, decls, lifetime, path, projectedOffs
             if Position.posGt r.End outerRange.Start then attrs else
             x.SkipOuterAttrs(rest, outerRange)
 
-    member x.ProcessTopLevelBinding(binding) =
-        let (SynBinding(_, kind, _, _, attrs, _, _ , headPat, returnInfo, expr, _, _)) = binding
+    member x.ProcessTopLevelBinding(binding, isSecond) =
+        let (SynBinding(_, kind, _, _, attrs, XmlDoc xmlDoc, _ , headPat, returnInfo, expr, range, _)) = binding
 
-        let mark = x.Mark()
+        if isSecond then x.AdvanceToTokenOrRangeStart(FSharpTokenType.AND, if xmlDoc.IsEmpty then range else xmlDoc.Range)
```

Comment:
Isn't `if xmlDoc.IsEmpty then range else xmlDoc.Range` logic extracted to the util in other places? Why place it inline here? Is it needed at all given `x.MarkOrMarkXmlDocOwner` is called below?

Reply:
I will probably need to revisit this logic again, try to simplify it, and come back to this question later.

Reply:
> Isn't if xmlDoc.IsEmpty then range else xmlDoc.Range logic extracted to the util in other places?

yes we have `xmlDocOwnerStartRange` for this

> Is it needed at all given x.MarkOrMarkXmlDocOwner is called below? It takes xml range into account.

It seems that all the same, it is necessary. `and` should be included in the binding declaration. Since the XmlDoc can appear either before `and` or after, and we only have a range for the XmlDoc and for the declaration without `and`, we must first advance to to XmlDoc or `and`.

Reply:
I think this also requires a comment in the code